### DISCUSSION TITLE
Clarification on the generic type <in out ClientType...> Syntax in the ODataService Class

### DISCUSSION THREAD

**Discussion by tobikugel** [odata2ts/odata2ts#312] [Q&A] [answered]
Hello,
I’ve encountered an issue in my project related to the generic type syntax used in the code, for example, the following declaration:
`export declare class ODataService<in out ClientType extends ODataHttpClient>`
This syntax causes type errors in my TypeScript environment. As I'm familiar with standard TypeScript syntax, the use of _in_ and _out_ here is unfamiliar and appears non-standard.

Could you please help clarify the purpose of the in and out keywords in this context? Is there any specific preprocessor, tool, or TypeScript configuration that I should set up in my project to resolve these type errors?

**Comment by texttechne** [comment, marked as answer]
Hi @tobikugel,
yeah, `in` and `out` are variance annotations and thus touch the complicated topic of covariance, contravariance and bivariance. 
I'm unable to point to the TypeScript docu in this regard, but you'll find a lot of articles covering this topic. But it's standard TS, introduced in 4.7.

Why did we specify all services as bivariant? Actually it's only there, so TS won't compute this information, which breaks with larger EDMX files. See:
* [OData2ts issue](https://github.com/odata2ts/odata2ts/issues/246)
* [TypeScript issue](https://github.com/microsoft/TypeScript/issues/58439)

So much for the background why this has been introduced. Now to your problem: What kind of error do you get for this code? Do you have TS >= 4.7 installed?

**Reply by tobikugel** [reply]
Thank you for your quick answer!
A TS version issue was the problem in my case.
